server/src/main/java/org/opensearch/indices/IndicesService.java

```diff
@@ -622,7 +635,9 @@ public NodeIndicesStats stats(CommonStatsFlags flags) {
                     break;
             }
         }
-
+        if (optimizedNodesStatsEnabled) {
+            return new NodeIndicesStats(commonStats, statsByShard(this, flags), searchRequestStats, flags.getLevels());
```

Comment:
On a quick read of your PR, it looks like the optimization you are referring to is to compute stats on the respective remote nodes instead of doing it in coordinator node which essentially distributes the compute. If this is correct then and if as per your benchmarks it has helped then it is certainly a good optimization. I have few questions though :
1. Can you please tell me how the new `statsByIndex` is being used? 
I can see that it is a new structure being passed to coordinator but I am not sure how it is helping in lesser compute. 
2. Initial read of the overview gave a hint that you are trying to change the default REST api but that is not the case right? It isn't the API but the data passed by transport actions. Your intention is to filter it out on respective nodes instead of passing the whole bunch to coordinator. If this is true then can you tell me how exactly you are achieving this? Is there some code still pending to be pushed in this PR?

Reply:
Hey @vikasvb90 thanks for taking a look at the PR. 
1. In the current approach, we compute Index level stats if level is passed from the rest layer as `indices`. This computation of IndexStats is done on the coordinator node instead of the data node from where the stats are coming in. By introducing the statsByIndex in the class, I am not just making sure that if the level is `indices`, we are still pre-computing the Index Stats.
2. I did some rework on the default behavior - that was indeed my original intention but since we do not want a breaking change in 2.x, I am have introduced a transport level flag which can be enabled for utilising this feature instead. Did not publish this PR as it was still in the discussion stage.